File: ios/BudgetBuddy/Repositories/AuthRepository.swift
import Foundation
import FirebaseAuth

enum AuthResult {
    case success(userId: String, displayName: String, email: String)
    case failure(String)
}

final class AuthRepository {
    private let auth    = Auth.auth()
    private let session: SessionManager
    private let firestore: FirestoreRepository
    private let sync: SyncRepository

    init(session: SessionManager, firestore: FirestoreRepository, sync: SyncRepository) {
        self.session   = session
        self.firestore = firestore
        self.sync      = sync
    }

    var isLoggedIn: Bool { auth.currentUser != nil }
    var currentUserId: String? { auth.currentUser?.uid }

    func signIn(email: String, password: String) async -> AuthResult {
        do {
            let result = try await auth.signIn(withEmail: email.trimmingCharacters(in: .whitespaces),
                                               password: password)
            let user = result.user
            let name = user.displayName?.nilIfEmpty
                ?? session.displayName
                ?? email.components(separatedBy: "@").first
                ?? "User"
            session.userId      = user.uid
            session.displayName = name
            session.email       = user.email ?? email
            Task {
                await sync.syncFromFirestore(userId: user.uid)
                if let url = await firestore.getUserPhotoUrl(uid: user.uid) {
                    session.photoUrl = url
                }
            }
            return .success(userId: user.uid, displayName: name, email: user.email ?? email)
        } catch {
            return .failure(friendlyError(error))
        }
    }

    func signUp(email: String, password: String, displayName: String) async -> AuthResult {
        do {
            let result = try await auth.createUser(withEmail: email.trimmingCharacters(in: .whitespaces),
                                                    password: password)
            let user = result.user
            let req = user.createProfileChangeRequest()
            req.displayName = displayName.trimmingCharacters(in: .whitespaces)
            try await req.commitChanges()
            session.userId      = user.uid
            session.displayName = displayName.trimmingCharacters(in: .whitespaces)
            session.email       = user.email ?? email
            Task {
                try? await firestore.saveUserProfile(uid: user.uid,
                                                     displayName: session.displayName ?? "",
                                                     email: user.email ?? email)
            }
            return .success(userId: user.uid, displayName: session.displayName ?? "", email: user.email ?? email)
        } catch {
            return .failure(friendlyError(error))
        }
    }

    func sendPasswordReset(email: String) async -> Result<Void, Error> {
        do {
            try await auth.sendPasswordReset(withEmail: email.trimmingCharacters(in: .whitespaces))
            return .success(())
        } catch {
            return .failure(error)
        }
    }

    func updatePhotoUrl(userId: String, url: String) async {
        let req = auth.currentUser?.createProfileChangeRequest()
        req?.photoURL = URL(string: url)
        try? await req?.commitChanges()
        try? await firestore.saveUserProfile(uid: userId,
                                             displayName: session.displayName ?? "",
                                             email: session.email ?? "",
                                             photoUrl: url)
        session.photoUrl = url
    }

    func signOut() {
        try? auth.signOut()
        session.clear()
    }

    private func friendlyError(_ error: Error) -> String {
        let code = (error as NSError).code
        let msg  = error.localizedDescription
        if msg.contains("user-not-found") || msg.contains("no user record") {
            return "No account found with this email address."
        }
        if msg.contains("wrong-password") || msg.contains("invalid-credential") || code == 17004 {
            return "Incorrect email or password."
        }
        if msg.contains("email-already-in-use") { return "An account with this email already exists." }
        if msg.contains("weak-password")        { return "Password is too weak." }
        if msg.contains("network")              { return "Network error. Check your connection." }
        if msg.contains("too-many-requests")    { return "Too many attempts. Try again later." }
        return msg
    }
}

private extension String {
    var nilIfEmpty: String? { isEmpty ? nil : self }
}

File: android/app/src/main/java/com/budgetbuddy/data/repository/AuthRepository.kt
package com.budgetbuddy.data.repository

import android.net.Uri
import android.util.Log
import com.budgetbuddy.data.local.SessionManager
import com.google.firebase.auth.FirebaseAuth
import com.google.firebase.auth.UserProfileChangeRequest
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.tasks.await
import javax.inject.Inject
import javax.inject.Singleton

private const val TAG = "AuthRepository"

sealed class AuthResult {
    data class Success(val userId: String, val displayName: String, val email: String) : AuthResult()
    data class Error(val message: String) : AuthResult()
}

@Singleton
class AuthRepository @Inject constructor(
    private val firebaseAuth: FirebaseAuth,
    private val sessionManager: SessionManager,
    private val syncRepository: SyncRepository,
    private val firestoreRepository: FirestoreRepository
) {
    val isLoggedIn: Boolean get() = firebaseAuth.currentUser != null
    val currentUserId: String? get() = firebaseAuth.currentUser?.uid
    val currentDisplayName: String? get() = sessionManager.displayName ?: firebaseAuth.currentUser?.displayName
    val currentEmail: String? get() = firebaseAuth.currentUser?.email
    val currentPhotoUrl: String? get() = sessionManager.photoUrl ?: firebaseAuth.currentUser?.photoUrl?.toString()

    suspend fun signIn(email: String, password: String): AuthResult {
        return try {
            val result = firebaseAuth.signInWithEmailAndPassword(email.trim(), password).await()
            val user = result.user ?: return AuthResult.Error("Sign in failed. Please try again.")
            val displayName = user.displayName?.takeIf { it.isNotBlank() }
                ?: sessionManager.displayName
                ?: email.substringBefore("@")
            sessionManager.userId = user.uid
            sessionManager.displayName = displayName
            sessionManager.email = user.email ?: email
            sessionManager.loginTimestamp = System.currentTimeMillis()
            CoroutineScope(Dispatchers.IO).launch {
                syncRepository.syncToFirestore(user.uid)
                syncRepository.syncFromFirestore(user.uid)
                val photoUrl = firestoreRepository.getUserPhotoUrl(user.uid)
                    ?: user.photoUrl?.toString()
                if (photoUrl != null) sessionManager.photoUrl = photoUrl
            }
            AuthResult.Success(user.uid, displayName, user.email ?: email)
        } catch (e: Exception) {
            AuthResult.Error(friendlyError(e.message))
        }
    }

    suspend fun signUp(email: String, password: String, displayName: String): AuthResult {
        return try {
            val result = firebaseAuth.createUserWithEmailAndPassword(email.trim(), password).await()
            val user = result.user ?: return AuthResult.Error("Registration failed. Please try again.")
            val profileUpdates = UserProfileChangeRequest.Builder()
                .setDisplayName(displayName.trim())
                .build()
            user.updateProfile(profileUpdates).await()
            sessionManager.userId = user.uid
            sessionManager.displayName = displayName.trim()
            sessionManager.email = user.email ?: email
            sessionManager.loginTimestamp = System.currentTimeMillis()
            CoroutineScope(Dispatchers.IO).launch {
                firestoreRepository.saveUserProfile(user.uid, displayName.trim(), user.email ?: email)
            }
            AuthResult.Success(user.uid, displayName.trim(), user.email ?: email)
        } catch (e: Exception) {
            AuthResult.Error(friendlyError(e.message))
        }
    }

    suspend fun sendPasswordReset(email: String): Result<Unit> {
        return try {
            firebaseAuth.sendPasswordResetEmail(email.trim()).await()
            Result.success(Unit)
        } catch (e: Exception) {
            Result.failure(Exception(friendlyError(e.message)))
        }
    }

    suspend fun updatePhotoUrl(userId: String, photoUrl: String) {
        try {
            val profileUpdates = UserProfileChangeRequest.Builder()
                .setPhotoUri(Uri.parse(photoUrl))
                .build()
            firebaseAuth.currentUser?.updateProfile(profileUpdates)?.await()
            firestoreRepository.saveUserProfile(
                userId,
                sessionManager.displayName ?: "",
                sessionManager.email ?: "",
                photoUrl
            )
            sessionManager.photoUrl = photoUrl
        } catch (e: Exception) {
            Log.w(TAG, "updatePhotoUrl failed: ${e.message}")
        }
    }

    fun signOut() {
        firebaseAuth.signOut()
        sessionManager.clear()
    }

    private fun friendlyError(message: String?): String = when {
        message == null -> "An error occurred. Please try again."
        message.contains("no user record") || message.contains("user-not-found") ->
            "No account found with this email address."
        message.contains("wrong-password") || message.contains("invalid-credential") ->
            "Incorrect email or password."
        message.contains("email-already-in-use") ->
            "An account with this email already exists."
        message.contains("weak-password") ->
            "Password is too weak. Please choose a stronger password."
        message.contains("network") || message.contains("Network") ->
            "Network error. Please check your connection."
        message.contains("too-many-requests") ->
            "Too many attempts. Please try again later."
        else -> message
    }
}
